### PR TITLE
[Widgets] Handle display logic

### PR DESCRIPTION
**Pull request by mishig25**
Closes https://github.com/huggingface/hub-docs/issues/1010

### This is how it looks like, when an inference api is disabled for a model.

image

### This is how it looks like, when an inference api is disabled for a model BUT there is an example with output (#979).
Detailed description:
1. There will be message that says `However, you can still play with [examples that have an output](hf.co/docs/hub/models-widgets#example-outputs)`
2. All the input elements are disabled (you can't enter text, the compute button is greyed out)
3. As a user, you only see examples that have outputs. If in the readme, there was an example without output, it is being filtered out.
4. In the current implementation, when you click an example with an output, it tries to call an inference-api. This should be fixed with #1017

image

### How to test

```
cd js
npm ci
npm run dev -- --open
```
and the command above should open a webpage

### Order of merge operations
1. Merge #1023 
2. Then merge #1063 
3. Then merge #1019 (this PR)

### CODE REVIEW

**Comment by julien-c** [comment]
@mishig25 had you seen that Figma from @gary149 ? https://www.figma.com/file/MHIIYlEdHujRkX0BpI56GS/text-to-image-widget-update?type=design&node-id=532-702&mode=design&t=8fszRBMxCjcZ7dbA-0

Maybe we can huddle quickly with also @SBrandeis

File: js/src/lib/components/InferenceWidget/InferenceWidget.svelte [hunk 1]
```diff
@@ -93,5 +101,62 @@
...
+		{#if hasExampleWithOutput}
+			<span class="text-sm text-gray-500">
+				However, you can still play with
+				<a class="color-inherit underline" href="/docs/hub/models-widgets#example-outputs"
+					>examples that have an output</a
+				>.
+			</span>
+			<svelte:component this={WIDGET_COMPONENTS[model.pipeline_tag ?? ""]} {...widgetProps} isDisabled={true} />
+		{/if}
+	{/if}
```
Comment (julien-c) [inline js/src/lib/components/InferenceWidget/InferenceWidget.svelte:161]: i think you don't really have a choice and you need to move that logic from moon-landing to this repo, right?

File: js/src/lib/components/InferenceWidget/shared/WidgetQuickInput/WidgetQuickInput.svelte [hunk 2]
```diff
@@ -16,11 +17,12 @@
 		{placeholder}
 		required={true}
 		type="text"
-		disabled={isLoading}
+		disabled={isLoading || isDisabled}
 	/>
 	<WidgetSubmitBtn
```
Comment (SBrandeis) [inline js/src/lib/components/InferenceWidget/shared/WidgetQuickInput/WidgetQuickInput.svelte:22]: Let's hide the sumbit button when disabled, no?

File: js/src/lib/components/InferenceWidget/shared/WidgetTextarea/WidgetTextarea.svelte [hunk 3]
```diff
@@ -82,13 +83,11 @@
...
 		<span
-			class="{isLoading ? 'pointer-events-none' : ''} {label
-				? 'mt-1.5'
-				: ''} block w-full resize-y overflow-auto py-2 px-3 {size === 'small'
+			class="{label ? 'mt-1.5' : ''} block w-full resize-y overflow-auto py-2 px-3 {size === 'small'
 				? 'min-h-[42px]'
 				: 'min-h-[144px]'} inline-block max-h-[500px] whitespace-pre-wrap rounded-lg border border-gray-200 shadow-inner outline-none focus:shadow-inner focus:ring focus:ring-blue-200 dark:bg-gray-925"
 			role="textbox"
-			contenteditable
+			contenteditable={!isLoading && !isDisabled}
```
Comment (SBrandeis) [inline js/src/lib/components/InferenceWidget/shared/WidgetTextarea/WidgetTextarea.svelte:90]: i think this shouldn't take focus if no editable

More generally (unrelated to this PR), I think it shouldn't take focus nor be editable while tex generation is happening

File: js/src/lib/components/InferenceWidget/shared/helpers.ts [hunk 4]
```diff
@@ -11,7 +12,10 @@ export function getSearchParams(keys: string[]): string[] {
 }
 
 export function getDemoInputs(model: ModelData, keys: (number | string)[]): any[] {
```
Comment (SBrandeis) [inline js/src/lib/components/InferenceWidget/shared/helpers.ts:14]: I'm confused about the role of this function. I see it's called from concrete Widgets' `onMount`, but there is also a similar filtering in `WidgetWrapper`:

https://github.com/huggingface/hub-docs/blob/557b21cbca1a5aa5e49173eab84662270b7670d8/js/src/lib/components/InferenceWidget/shared/WidgetWrapper/WidgetWrapper.svelte#L41

It would help if the responsibility of preparing/filtering the examples & chosing which one to display first was done all in the same place.

File: js/src/lib/components/InferenceWidget/widgets/AudioToAudioWidget/AudioToAudioWidget.svelte [hunk 5]
```diff
@@ -168,7 +170,7 @@
 >
 	<svelte:fragment slot="top">
 		<form>
-			<div class="flex flex-wrap items-center">
+			<div class="flex flex-wrap items-center {isDisabled ? 'pointer-events-none opacity-50' : ''}">
```
Comment (SBrandeis) [inline js/src/lib/components/InferenceWidget/widgets/AudioToAudioWidget/AudioToAudioWidget.svelte:173]: I think we should also hide input controls as well when disabled

@julien-c do you have an opinion?

File: js/src/lib/components/InferenceWidget/InferenceWidget.svelte [hunk 6]
```diff
@@ -93,5 +101,62 @@
...
+			</span>
+		{/if}
+
+		<!-- If there is an example with output, then show the widget with disabled interaction -->
+		{#if hasExampleWithOutput}
+			<span class="text-sm text-gray-500">
+				However, you can still play with
+				<a class="color-inherit underline" href="/docs/hub/models-widgets#example-outputs"
+					>examples that have an output</a
+				>.
```
Comment (julien-c) [inline js/src/lib/components/InferenceWidget/InferenceWidget.svelte:157]: no need to write something so specific IMO, but what we had discussed was changing the `Inference API` header to `Inference Examples` in that case.

cc @gary149 still agree?

File: js/src/lib/components/InferenceWidget/shared/helpers.ts (hunk 4 above)
Comment (julien-c) [reply js/src/lib/components/InferenceWidget/shared/helpers.ts:14]: related to #1021 ?

File: js/src/lib/components/InferenceWidget/widgets/AudioToAudioWidget/AudioToAudioWidget.svelte (hunk 5 above)
Comment (julien-c) [reply js/src/lib/components/InferenceWidget/widgets/AudioToAudioWidget/AudioToAudioWidget.svelte:173]: yes that was the spirit of @gary149's Figma i think

Comment (SBrandeis) [reply js/src/lib/components/InferenceWidget/widgets/AudioToAudioWidget/AudioToAudioWidget.svelte:173]: Ah just remembered this is included in the Figma

File: js/src/lib/components/InferenceWidget/InferenceWidget.svelte (hunk 1 above)
Comment (mishig25) [reply js/src/lib/components/InferenceWidget/InferenceWidget.svelte:161]: yes, otherwise, the widget was completely hidden from moon-landing. But we want to decide whether to hide or not based on available examples with outputs

File: js/src/lib/components/InferenceWidget/InferenceWidget.svelte (hunk 6 above)
Comment (gary149) [reply js/src/lib/components/InferenceWidget/InferenceWidget.svelte:157]: yes

File: js/src/lib/components/InferenceWidget/shared/helpers.ts (hunk 4 above)
Comment (SBrandeis) [reply js/src/lib/components/InferenceWidget/shared/helpers.ts:14]: yes!
Let's go for #1021 and rely on the `WidgetWrapper` to filter examples then?
Wdyt @mishig25 ?

Comment (julien-c) [reply js/src/lib/components/InferenceWidget/shared/helpers.ts:14]: yes! ^^

File: js/src/lib/components/InferenceWidget/widgets/AudioToAudioWidget/AudioToAudioWidget.svelte (hunk 5 above)
Comment (gary149) [reply js/src/lib/components/InferenceWidget/widgets/AudioToAudioWidget/AudioToAudioWidget.svelte:173]: > I think we should also hide input controls as well when disabled

Yes.

File: js/src/lib/components/InferenceWidget/InferenceWidget.svelte [hunk 7]
```diff
@@ -93,5 +101,62 @@
...
+			<span class="text-sm text-gray-500">
+				Inference API is disabled for an unknown reason. Please open a
+				<a class="color-inherit underline" href="/{model.id}/discussions/new">Discussion in the Community tab</a>.
+			</span>
+		{/if}
+
+		<!-- If there is an example with output, then show the widget with disabled interaction -->
+		{#if hasExampleWithOutput}
+			<span class="text-sm text-gray-500">
+				However, you can still play with
```
Comment (Pierrci) [inline js/src/lib/components/InferenceWidget/InferenceWidget.svelte:154]:
```suggestion
				You can still play with
```
too wordy for my taste otherwise, when combined with the text above :)

File: js/src/lib/components/InferenceWidget/shared/helpers.ts (hunk 4 above)
Comment (mishig25) [reply js/src/lib/components/InferenceWidget/shared/helpers.ts:14]: handled by https://github.com/huggingface/hub-docs/pull/1023

**Comment by julien-c** [comment]
this is ready to merge no?

**Comment by mishig25** [comment]
I need to merge https://github.com/huggingface/hub-docs/pull/1063 first

File: js/src/lib/components/InferenceWidget/InferenceWidget.svelte [hunk 8]
```diff
@@ -93,5 +101,62 @@
...
+				<a class="color-inherit underline" href="/docs/hub/models-widgets#example-outputs"
+					>examples that have an output</a
+				>.
+			</span>
+			<svelte:component this={WIDGET_COMPONENTS[model.pipeline_tag ?? ""]} {...widgetProps} isDisabled={true} />
```
Comment (SBrandeis) [inline js/src/lib/components/InferenceWidget/InferenceWidget.svelte:159]: The examples are not filtered, meaning you also get examples without an output here

File: js/src/lib/components/InferenceWidget/InferenceWidget.svelte [hunk 9]
```diff
@@ -93,5 +101,62 @@
...
+		{:else}
+			<!-- added as a failsafe but this case cannot currently happen -->
+			<span class="text-sm text-gray-500">
+				Inference API is disabled for an unknown reason. Please open a
+				<a class="color-inherit underline" href="/{model.id}/discussions/new">Discussion in the Community tab</a>.
+			</span>
+		{/if}
+
+		<!-- If there is an example with output, then show the widget with disabled interaction -->
+		{#if hasExampleWithOutput}
```
Comment (SBrandeis) [inline js/src/lib/components/InferenceWidget/InferenceWidget.svelte:152]: Only a few modalities support outputs in widget examples.

Nothing prevents me from setting an example with output in the metadata of a aduio2audio; I believe in this case the behavior of the widget is to try to call the Inference API, which will fail most likely

File: js/src/lib/components/InferenceWidget/shared/WidgetQuickInput/WidgetQuickInput.svelte (hunk 2 above)
Comment (mishig25) [reply js/src/lib/components/InferenceWidget/shared/WidgetQuickInput/WidgetQuickInput.svelte:22]: same comment as here https://github.com/huggingface/hub-docs/pull/1019#discussion_r1358067690

**Review by SBrandeis — COMMENTED** [review]
LGTM overall, I have two concerns about this PR (see comments)

The factorization of WidgetWrapper / InferenceWidget components is nice 👌 

I regret that we have to edit so many files to pass down `isDisabled` everywhere :/

File: js/src/lib/components/InferenceWidget/InferenceWidget.svelte (hunk 8 above)
Comment (mishig25) [reply js/src/lib/components/InferenceWidget/InferenceWidget.svelte]: handled in https://github.com/huggingface/hub-docs/pull/1019/commits/e091361aec6fbe914efc292f8888da858bd4e2b1

File: js/src/lib/components/InferenceWidget/InferenceWidget.svelte (hunk 9 above)
Comment (mishig25) [reply js/src/lib/components/InferenceWidget/InferenceWidget.svelte]: yes, this will be handled in subseq PR as discussed in https://github.com/huggingface/hub-docs/pull/1063#discussion_r1377293060

**Comment by mishig25** [comment]
### Update the UI to look like the [figma spec](https://www.figma.com/file/MHIIYlEdHujRkX0BpI56GS/text-to-image-widget-update?type=design&node-id=532-702&mode=design&t=8fszRBMxCjcZ7dbA-0)

Screenshot 2023-10-31 at 13 59 50

When widget is disabled but with examples outputs:
1. Compute btn is hidden
2. Add file btn is hidden
3. JSON output btn is hidden
4. Text areas don't have placeholders and unineractive
5. Fiel uploads don't have placeholders and uninteractive

Screenshot 2023-10-31 at 13 59 59

Screenshot 2023-10-31 at 14 01 48

File: js/src/lib/components/InferenceWidget/InferenceWidget.svelte (hunk 6 above)
Comment (mishig25) [reply js/src/lib/components/InferenceWidget/InferenceWidget.svelte]: see updated screenshot [here](https://github.com/huggingface/hub-docs/pull/1019#issuecomment-1787280284)

File: js/src/lib/components/InferenceWidget/widgets/AudioToAudioWidget/AudioToAudioWidget.svelte (hunk 5 above)
Comment (mishig25) [reply js/src/lib/components/InferenceWidget/widgets/AudioToAudioWidget/AudioToAudioWidget.svelte]: handled in [here](https://github.com/huggingface/hub-docs/pull/1019#issuecomment-1787280284)

File: js/src/lib/components/InferenceWidget/shared/WidgetWrapper/WidgetWrapper.svelte [hunk 10]
```diff
@@ -33,39 +34,57 @@
 	export let applyInputSample: (sample: TWidgetExample, opts?: ExampleRunOpts) => void = () => {};
 	export let validateExample: (sample: WidgetExample) => sample is TWidgetExample;
 	export let exampleQueryParams: QueryParam[] = [];
+	export let isDisabled = false;
+	isDisabled = model.inference !== InferenceDisplayability.Yes && model.pipeline_tag !== "reinforcement-learning";
```
Comment (SBrandeis) [inline js/src/lib/components/InferenceWidget/shared/WidgetWrapper/WidgetWrapper.svelte:38]: Should this statement be reactive?

File: js/src/lib/components/InferenceWidget/shared/WidgetWrapper/WidgetWrapper.svelte [hunk 11]
```diff
@@ -33,39 +34,57 @@
 	export let applyInputSample: (sample: TWidgetExample, opts?: ExampleRunOpts) => void = () => {};
 	export let validateExample: (sample: WidgetExample) => sample is TWidgetExample;
 	export let exampleQueryParams: QueryParam[] = [];
+	export let isDisabled = false;
```
Comment (SBrandeis) [inline js/src/lib/components/InferenceWidget/shared/WidgetWrapper/WidgetWrapper.svelte:37]: I think it gets overridden directly by the line below no?
```suggestion
```

File: js/src/lib/components/InferenceWidget/widgets/AudioClassificationWidget/AudioClassificationWidget.svelte [hunk 12]
```diff
@@ -163,6 +164,7 @@
 	{computeTime}
 	{error}
 	{isLoading}
+	bind:isDisabled
```
Comment (SBrandeis) [inline js/src/lib/components/InferenceWidget/widgets/AudioClassificationWidget/AudioClassificationWidget.svelte:167]: You should use `let:isDisabled` I think

https://svelte.dev/docs/special-elements#slot-slot-key-value

File: js/src/lib/components/InferenceWidget/shared/WidgetWrapper/WidgetWrapper.svelte (hunk 11 above)
Comment (SBrandeis) [reply js/src/lib/components/InferenceWidget/shared/WidgetWrapper/WidgetWrapper.svelte:37]: Ah after re-reading I'm thinking this should probably not be exported
```suggestion
	let isDisabled = false;
```

File: js/src/lib/components/InferenceWidget/widgets/AudioClassificationWidget/AudioClassificationWidget.svelte (hunk 12 above)
Comment (mishig25) [reply js/src/lib/components/InferenceWidget/widgets/AudioClassificationWidget/AudioClassificationWidget.svelte]: great suggestion. Makes the code cleaner c6bd915804c0459aca9e6916e58cadd59b771f37

File: js/src/lib/components/InferenceWidget/shared/WidgetWrapper/WidgetWrapper.svelte [hunk 13]
```diff
@@ -33,39 +34,57 @@
...
+	let inputSamples = !isDisabled ? inputSamplesAll : inputSamplesAll.filter(sample => sample.output !== undefined);
+	let inputGroups = getExamplesGroups();
 
 	$: selectedInputSamples =
 		inputGroups.length === 1 ? inputGroups[0] : inputGroups.find(({ group }) => group === selectedInputGroup);
 
+	function getExamplesGroups(): ExamplesGroup[] {
+		const inputGroups: ExamplesGroup[] = [];
+		for (const inputSample of inputSamples) {
+			const isExist = inputGroups.find(({ group }) => group === inputSample.group);
```
[the remark this answers is not in the document]
Comment (mishig25) [reply js/src/lib/components/InferenceWidget/shared/WidgetWrapper/WidgetWrapper.svelte]: 5ae1084e61848caa96a7ae73a009bbe92ddae627

File: js/src/lib/components/InferenceWidget/shared/WidgetWrapper/WidgetWrapper.svelte [hunk 14]
```diff
@@ -33,39 +34,57 @@
...
 	let selectedInputGroup: string;
+	let modelTooBig = false;
 
-	const inputSamples = (model.widgetData ?? [])
+	interface ExamplesGroup {
+		group: string;
+		inputSamples: TWidgetExample[];
+	}
+
+	const inputSamplesAll = (model.widgetData ?? [])
```
[the remark this answers is not in the document]
Comment (mishig25) [reply js/src/lib/components/InferenceWidget/shared/WidgetWrapper/WidgetWrapper.svelte]: e701e955fa961fc2b80f422edd719e59deee3e70

File: js/src/lib/components/InferenceWidget/shared/WidgetWrapper/WidgetWrapper.svelte (hunk 10 above)
Comment (mishig25) [reply js/src/lib/components/InferenceWidget/shared/WidgetWrapper/WidgetWrapper.svelte]: it is being changed inside onMount https://github.com/huggingface/hub-docs/blob/c6bd915804c0459aca9e6916e58cadd59b771f37/js/src/lib/components/InferenceWidget/shared/WidgetWrapper/WidgetWrapper.svelte#L81-L83

otherwise, there is no need for `$: ` svelte syntax

**Comment by julien-c** [comment]
let's merge the hub-docs update in moon-landing?
